### ISSUE TITLE
TBaseVirtualTree.GetLastVisibleNoInit code is incorrect and causes exponential slowdowns when traversing very large trees. (Suggested code changes attached)

### ISSUE THREAD

**Issue by eclipse666** [JAM-Software/Virtual-TreeView#725]
I've been using TreeView 4.5.4 with a Delphi 2007 project which I recently ported across to Delphi 10.2
In the process I installed the latest TreeView 6.7.0
I noticed that my TreeView control was very sluggish when moving up or down nodes using the arrow keys and there was also a lag when clicking on items far down the tree. This did not happen with the 4.5.4 version. 

I've traced the problem to the TBaseVirtualTree.GetLastVisibleNoInit function:

6.7.0 looks like this:

```
function TBaseVirtualTree.GetLastVisibleNoInit(Node: PVirtualNode = nil;
  ConsiderChildrenAbove: Boolean = True; IncludeFiltered: Boolean = False): PVirtualNode;
 
// Returns the very last visible node in the tree while optionally considering toChildrenAbove.
// No initialization is performed.

begin
  Result := GetLastNoInit(Node, ConsiderChildrenAbove);
  while Assigned(Result) and (Result <> Node) do
  begin
    if FullyVisible[Result] and
       (IncludeFiltered or not IsEffectivelyFiltered[Result]) then
      Break;
    Result := GetPreviousNoInit(Result, ConsiderChildrenAbove);
  end;

  if (Result = Node) then // i.e. there is no visible node
    Result := nil;
end;
```

The same function in 4.5.4 looks like this:

```
function TBaseVirtualTree.GetLastVisibleNoInit(Node: PVirtualNode = nil): PVirtualNode;
 
// Returns the very last visible node in the tree without initialization.
 
var
  Next: PVirtualNode;
 
begin
  Result := GetLastVisibleChildNoInit(Node);
  while Assigned(Result) do
  begin
    // Test if there is a next last visible child. If not keep the node from the last run.
    // Otherwise use the next last visible child.
    Next := GetLastVisibleChildNoInit(Result);
    if Next = nil then
      Break;
    Result := Next;
  end;
end;
```

As you can see this version is calling the GetLastVisibleChildNoInit function instead of GetLastNoInit and GetPreviousNoInit

My suggested change is below:

```
function TBaseVirtualTree.GetLastVisibleNoInit(Node: PVirtualNode = nil;
  ConsiderChildrenAbove: Boolean = True; IncludeFiltered: Boolean = False): PVirtualNode;

// Returns the very last visible node in the tree while optionally considering toChildrenAbove.
// No initialization is performed.

begin
  Result := GetLastVisibleChildNoInit(Node, ConsiderChildrenAbove);
  while Assigned(Result) and (Result <> Node) do
  begin
    if FullyVisible[Result] and
       (IncludeFiltered or not IsEffectivelyFiltered[Result]) then
      Break;

    Result := GetLastVisibleChildNoInit(Result, ConsiderChildrenAbove);
  end;

  if (Result = Node) then // i.e. there is no visible node
    Result := nil;
end;

```
Making this change resolved the sluggishness I was experiencing in my project.

**Comment by joachimmarder** [comment]
In the change-log for V5.0.0RC2 there is a note about a change regarding `GetLastVisibleNoInit()`:
```
Bug fix: Functions GetLastVisible and GetLastVisibleNoInit return correct results even if some anchestor of the last visible node is not effectively visible
```
Unfortunately there is no issue mentioned for this change. I didn't do anything in this part of the code since I started maintaining this project. Can you describe why your change is faster? Willit alter the behavior of V5/V6 or might it bring back the problem resolved in V5.0.0RC2?

**Comment by eclipse666** [comment]
My app uses a TVirtualDrawTree that loads an entire hard drive's directory and file structure (similar to TreeSize). The tree is sorted by size with largest folder and file sizes appearing at the top. There's a single root node for the hard drive that's selected (e.g. C:\) and all the sub folders hang off it.

I noticed that the further a node was from the top of the tree the longer it took for any clicks or up/down movements to be processed. Near the bottom of the tree I was waiting for up to 1 second from the time I clicked it to the time it displayed itself as selected. I'd have the same delay when using the up or down arrow to move the selection. 

I put code into various tree view functions to count how many times they were triggered. I found that the GetLastNoInit function was being called anything from hundreds of thousands to millions of times for every individual click. The number of calls increased exponentially with the distance to the root node.

Interestingly enough enabling the "ChildrenAbove" option in the tree made the problem go away but that wasn't a solution for me.

I knew this wasn't a problem in the older 4.5.4 version so I compared the 2 and made the changes I described. The number of calls to GetLastNoInit dropped to no more than a few hundred per click after that.

I'm not sure if this change will cause any other problems but the existing code is certainly incorrect if it's triggering millions of calls for each click.

**Comment by joachimmarder** [comment]
> I found that the `GetLastNoInit` function was being called anything from hundreds of thousands to millions of times for every individual click. The number of calls increased exponentially with the distance to the root node.

Well, that indeed sounds bad. Interestingly I never recognized this in TreeSize.

Are you able to create a sample project that allows to replicate the problem? Maybe it could be based on one of the sample apps than come with Virtual TreeView. It could be interesting to analyze the call stacks that lead to the calls of `GetLastNoInit()`. I personally never look into that part of the code, but I need to make sure that this change does not cause other problem to (re-)occur.

**Comment by eclipse666** [comment]
Modify the "Advanced" demo as follows

VirtualTrees.pas:

```
var // Clipboard format IDs used in OLE drag'n drop and clipboard transfers.
  CF_VIRTUALTREE,
  CF_VTREFERENCE,
  CF_VRTF,
  CF_VRTFNOOBJS,   // Unfortunately CF_RTF* is already defined as being
                   // registration strings so I have to use different identifiers.
  CF_HTML,
  CF_CSV: Word;

  MMXAvailable: Boolean; // necessary to know because the blend code uses MMX instructions
  IsWinVistaOrAbove: Boolean;

  {$MinEnumSize 1, make enumerations as small as possible}
  glni : longint; // <----------- add this line

...

function TBaseVirtualTree.GetLastNoInit(Node: PVirtualNode = nil; ConsiderChildrenAbove: Boolean = False): PVirtualNode;

// Returns the very last node in the tree branch given by Node without initialization.

var
  Next: PVirtualNode;

begin
  inc( glni ); // <-------------- add this line

  Result := GetLastChildNoInit(Node);
  if not ConsiderChildrenAbove or not (toChildrenAbove in FOptions.FPaintOptions) then
    while Assigned(Result) do
    begin
      // Test if there is a next last child. If not keep the node from the last run.
      // Otherwise use the next last child.
      Next := GetLastChildNoInit(Result);
      if Next = nil then
        Break;
      Result := Next;
    end;
end;
```  

SpeedDemo.pas:
Add labelCount to top area of form (to display GetLastNoInit count and change since last call)
Add timer1 to form, set to 10ms interval

```
procedure TSpeedForm.Timer1Timer(Sender: TObject);

begin
  if labelCount.Tag <> glni then
  begin
    labelCount.Caption := Format( '%d (+%d)', [ glni, glni - labelCount.Tag ] );
    labelCount.Tag := glni;
  end;
end;

```

In speed demo:
select 100000 nodes
Click add nodes to root

Select topmost node (or any node near top), then click "add nodes as children"
Move to bottom of the tree. Click or navigate near the bottom
You'll see extremely large numbers with the original code. Numbers are highest at the bottom of the tree.

**Comment by joachimmarder** [comment]
@DanielBauten: The code change was done by you in commit 3cd3660 on 11 June 2012. Can you please comment on the suggested change?

**Comment by DanielBauten** [comment]
The problem with sluggishness should only occur if there are many non-visible nodes at the end of the subtree of `Node`, which causes `GetPreviousNoInit` to be called very frequently.

At first sight, however, it seems to me that the suggested change

1. **brings back** the **problem solved in V5.0.0RC2** (in case _"if some anchestor of the last visible node is not effective visible"_) and 
2. no further levels are taken into account in case that the last direct child is visible (since `Break` is called to early) – i.e. the **result** is therefore **incorrect** if the depth of the subtree of `Node` is greater than 2.

@eclipse666: Can you revise your proposal?

**Comment by ElminsterAU** [comment]
I'm using Virtual TreeView as primary UI element in https://github.com/TES5Edit/TES5Edit  
I've recently updated the (significantly modified, available at https://github.com/eFrysTon/Virtual-TreeView ) used code from something ancient to the latest development version from here.

I too have noticed the VERY significant slowdown which makes Virtual TreeView close to unusable. (About 2 seconds to move one node when pressing cursor down while looking at the end of the tree).

I think some form of fix for this is very much needed.

**Comment by joachimmarder** [comment]
It would be great if some could post a simple sample project here.

**Comment by joachimmarder** [comment]
@ElminsterAU : Would you say [your fix](https://github.com/ElminsterAU/Virtual-TreeView/commit/2064c7c4bc8bc63e75133e2c871bff821bad2fd7) is a good and general solution for the problem? I would then apply it to the Virtual TreeView master.

**Comment by ElminsterAU** [comment]
@joachimmarder : no, unfortunately not. I reverted that "fix" (which I got from somewhere else) again as it caused problems.

But looking at the code with fresh eyes again now, the issue with the existing code is that GetLastNoInit and GetPreviousNoInit which are being called here go deeply down branches even if a parent further up the branch already isn't FullyVisible or is filtered, wasting a lot of time.

The solution I can see here is to copy the code of GetLastNoInit and replace the GetLastChildNoInit calls with GetLastVisibleChildNoInit calls. For better performance the 2nd GetLastVisibleChildNoInit call can be replaced with some code extracted from GetLastVisibleChildNoInit, to prevent unnecessary calls to FullyVisible.

Initial testing seems to indicate that this solves the performance problem I was seeing while not causing any of the issues of the previous fix (that is, as far as I can tell, the function will continue to return exactly the same results for the same inputs, just faster now).

The new commit is referenced above.
